### PR TITLE
[Feature]: Add global silent param for ConversableAgent

### PR DESCRIPTION
**Pull request by wenngong**
## Why are these changes needed?
Add a silent param to globally silent one agent.

## Related issue number
#3242

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by umermansoor** [comment]
Question: how does this relate to the `silent` param in the `initiate_chat` method? 
I believe the logic will override `initiate_chat`, meaning if I pass `silent=False` in `initiate_chat` and then do `is_silent=False` for any one of the agents, all agents will be unmuted.

```
def initiate_chat(
        self,
        recipient: "ConversableAgent",
        clear_history: bool = True,
        silent: Optional[bool] = False,
        cache: Optional[AbstractCache] = None,
        max_turns: Optional[int] = None,
        summary_method: Optional[Union[str, Callable]] = DEFAULT_SUMMARY_METHOD,
        summary_args: Optional[dict] = {},
        message: Optional[Union[Dict, str, Callable]] = None,
        **kwargs,
    )
   ```

File: autogen/agentchat/conversable_agent.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def __init__(
         default_auto_reply: Union[str, Dict] = "",
         description: Optional[str] = None,
         chat_messages: Optional[Dict[Agent, List[Dict]]] = None,
+        is_silent: Optional[bool] = None,
```
Comment (umermansoor) [inline autogen/agentchat/conversable_agent.py:81]: should this be called `silent` for consistency with `initiate_chat`?

Comment (wenngong) [reply autogen/agentchat/conversable_agent.py]: @umermansoor should be OK now?

**Comment by wenngong** [comment]
> Question: how does this relate to the `silent` param in the `initiate_chat` method? I believe the logic will override `initiate_chat`, meaning if I pass `silent=False` in `initiate_chat` and then do `is_silent=False` for any one of the agents, all agents will be unmuted. This will lead to confusing behavior
> 
> ```
> def initiate_chat(
>         self,
>         recipient: "ConversableAgent",
>         clear_history: bool = True,
>         silent: Optional[bool] = False,
>         cache: Optional[AbstractCache] = None,
>         max_turns: Optional[int] = None,
>         summary_method: Optional[Union[str, Callable]] = DEFAULT_SUMMARY_METHOD,
>         summary_args: Optional[dict] = {},
>         message: Optional[Union[Dict, str, Callable]] = None,
>         **kwargs,
>     )
> ```

@umermansoor As the origin feature requested in #3242, agent's silent setting will override any other silent settings include initiate_chat silent param. As below function I sumbmitted:
```
def _is_silent(self, silent: Optional[bool] = False) -> bool:
        return self.silent if self.silent is not None else silent
```

**Comment by Hk669** [comment]
@wenngong can you please verify these changes with the groupchat and let us know.

File: autogen/agentchat/conversable_agent.py [hunk 2]
```diff
@@ -604,6 +611,8 @@ def _process_message_before_send(
         self, message: Union[Dict, str], recipient: Agent, silent: bool
     ) -> Union[Dict, str]:
         """Process the message before sending it to the recipient."""
+        silent = self._is_silent(silent)
+
         hook_list = self.hook_lists["process_message_before_send"]
         for hook in hook_list:
             message = hook(sender=self, message=message, recipient=recipient, silent=silent)
```
Comment (thinkall) [inline autogen/agentchat/conversable_agent.py:618]:
```suggestion
        hook_list = self.hook_lists["process_message_before_send"]
        for hook in hook_list:
            message = hook(sender=self, message=message, recipient=recipient, silent=self._is_silent(silent))
```

File: autogen/agentchat/conversable_agent.py [hunk 3]
```diff
@@ -648,6 +657,8 @@ def send(
         Raises:
             ValueError: if the message can't be converted into a valid ChatCompletion message.
         """
+        silent = self._is_silent(silent)
+
         message = self._process_message_before_send(message, recipient, silent)
```
Comment (thinkall) [inline autogen/agentchat/conversable_agent.py:662]:
```suggestion
        message = self._process_message_before_send(message, recipient, self._is_silent(silent))
```

**Review by thinkall — COMMENTED** [review]
Thank you for the PR, @wenngong !

In the current implementation, the single agent's `silent` will overwrite the overall chat's silent setting. But it should only overwrite its own print behavior. I've left two suggestions as examples.

**Comment by wenngong** [comment]
> Thank you for the PR, @wenngong !
> 
> In the current implementation, the single agent's `silent` will overwrite the overall chat's silent setting. But it should only overwrite its own print behavior. I've left two suggestions as examples.

@thinkall You are wright! Since 'silent' passed by init_chat will transfer between agents, self._is_silent(silent) should only affect the agent's local print behavior, ant keep the origin 'silent' param transfer between agents as it was.

File: autogen/agentchat/conversable_agent.py (hunk 3 above)
Comment (wenngong) [reply autogen/agentchat/conversable_agent.py]: updated!

File: autogen/agentchat/conversable_agent.py (hunk 2 above)
Comment (wenngong) [reply autogen/agentchat/conversable_agent.py]: updated!

**Comment by wenngong** [comment]
> @wenngong can you please verify these changes with the groupchat and let us know.

Hi @Hk669 , 

in groupchat, GroupChatManager always set `silent=True` when broadcast msg to agents (groupchat.py  line 1041). 
So when all agents' silent param do not set (leave to default None), the example codes and outputs are:
```
from autogen import AssistantAgent, UserProxyAgent, config_list_from_json, ConversableAgent, runtime_logging
from autogen import GroupChat, GroupChatManager

config_list = config_list_from_json(env_or_file="OAI_CONFIG_LIST")

# The Number Agent always returns the same numbers.
number_agent = ConversableAgent(
    name="Number_Agent",
    system_message="You return me the numbers I give you, one number each line.",
    llm_config={"config_list": config_list},
    human_input_mode="NEVER",
)

# The Adder Agent adds 1 to each number it receives.
adder_agent = ConversableAgent(
    name="Adder_Agent",
    system_message="You add 1 to each number I give you and return me the new numbers, one number each line.",
    llm_config={"config_list": config_list},
    human_input_mode="NEVER",
)

# The Subtracter Agent subtracts 1 from each number it receives.
subtracter_agent = ConversableAgent(
    name="Subtracter_Agent",
    system_message="You subtract 1 from each number I give you and return me the new numbers, one number each line.",
    llm_config={"config_list": config_list},
    human_input_mode="NEVER",
)

# The `description` attribute is a string that describes the agent.
# It can also be set in `ConversableAgent` constructor.
adder_agent.description = "Add 1 to each input number."
subtracter_agent.description = "Subtract 1 from each input number."
number_agent.description = "Return the numbers given."

group_chat = GroupChat(
    agents=[adder_agent, subtracter_agent, number_agent],
    messages=[],
    max_round=6,
)

group_chat_manager = GroupChatManager(
    groupchat=group_chat,
    llm_config={"config_list": config_list},
)

chat_result = number_agent.initiate_chat(
    group_chat_manager,
    message="My number is 3, I want to turn it into 13.",
    summary_method="reflection_with_llm",
)
```

```
# outputs:

Number_Agent (to chat_manager):

My number is 3, I want to turn it into 13.

--------------------------------------------------------------------------------

Next speaker: Adder_Agent

Adder_Agent (to chat_manager):

4

--------------------------------------------------------------------------------

Next speaker: Subtracter_Agent

Subtracter_Agent (to chat_manager):

3

--------------------------------------------------------------------------------
/Users/vinnie/Code/ai/autogen/autogen/oai/anthropic.py:420: UserWarning: Cost calculation not available for model anthropic.claude-v2:1
  warnings.warn(f"Cost calculation not available for model {model}", UserWarning)

Next speaker: Subtracter_Agent

Subtracter_Agent (to chat_manager):

Unfortunately I don't have enough context to continue generating numbers for you. If you provide me another number, I'd be happy to subtract 1 from it and return the new number. Please let me know if you'd like me to continue this process!

--------------------------------------------------------------------------------

Next speaker: Adder_Agent

Adder_Agent (to chat_manager):

Sure, here is another number: 7

--------------------------------------------------------------------------------

Next speaker: Subtracter_Agent
```

Then, set adder_agent `silent=False` with other code keep the same：
```
# The Adder Agent adds 1 to each number it receives.
adder_agent = ConversableAgent(
    name="Adder_Agent",
    system_message="You add 1 to each number I give you and return me the new numbers, one number each line.",
    llm_config={"config_list": config_list},
    human_input_mode="NEVER",
    silent=False,
)
```

```
# outputs:

Number_Agent (to chat_manager):

My number is 3, I want to turn it into 13.

--------------------------------------------------------------------------------
chat_manager (to Adder_Agent):

My number is 3, I want to turn it into 13.

--------------------------------------------------------------------------------

Next speaker: Adder_Agent

Adder_Agent (to chat_manager):

4

--------------------------------------------------------------------------------

Next speaker: Subtracter_Agent

Subtracter_Agent (to chat_manager):

3

--------------------------------------------------------------------------------
chat_manager (to Adder_Agent):

3

--------------------------------------------------------------------------------

Next speaker: Subtracter_Agent

Subtracter_Agent (to chat_manager):

Unfortunately I don't have enough context to continue generating numbers for you. If you provide me another number, I'd be happy to subtract 1 from it and return the new number. Please let me know if you'd like me to continue this process!
```

As we can see above, when set adder_agent `silent=False`, adder_agent will print verbose log (in _print_received_message func, conversable_agent.py line 722):

```
--------------------------------------------------------------------------------
chat_manager (to Adder_Agent):

My number is 3, I want to turn it into 13.
```

**Comment by umermansoor** [comment]
@wenngong Thanks for making the changes requested. I still don't think it's working properly. See below where I'm trying to silence _Alice_.

```
alice = ConversableAgent(
        "alice",
        system_message="You are pretending to be Alice",
        human_input_mode="NEVER",
        llm_config={"config_list": config_list},
        silent=True,
)

charlie = ConversableAgent(
    "charlie",
    system_message="You are pretending to be Charlie",
    human_input_mode="NEVER",
    llm_config={"config_list": config_list}
)
alice.initiate_chat(
        charlie,
        message=f"How are you doing? What's your name.",
        max_turns=2,
        clear_history=False,
)
```

output:

```
alice (to charlie):

How are you doing? What's your name.

--------------------------------------------------------------------------------
alice (to charlie):

Nice to meet you too, Charlie! Since I'm pretending to be Alice, I'm going to continue the conversation in character. 

As Alice, I might be looking for some information on baking a cake or need help fixing a computer issue. Or perhaps I'm in a mood for a good book recommendation or just a friendly chat. Is there anything particular you'd want to know or discuss with Alice?
```

**Comment by wenngong** [comment]
> @wenngong Thanks for making the changes requested. I still don't think it's working properly. See below where I'm trying to silence _Alice_. It mutes Charlie (the other agent instead.)
> 
> ```
> alice = ConversableAgent(
>         "alice",
>         system_message="You are pretending to be Alice",
>         human_input_mode="NEVER",
>         llm_config={"config_list": config_list},
>         silent=True,
> )
> 
> charlie = ConversableAgent(
>     "charlie",
>     system_message="You are pretending to be Charlie",
>     human_input_mode="NEVER",
>     llm_config={"config_list": config_list}
> )
> alice.initiate_chat(
>         charlie,
>         message=f"How are you doing? What's your name.",
>         max_turns=2,
>         clear_history=False,
> )
> ```
> 
> output:
> 
> ```
> alice (to charlie):
> 
> How are you doing? What's your name.
> 
> --------------------------------------------------------------------------------
> alice (to charlie):
> 
> Nice to meet you too, Charlie! Since I'm pretending to be Alice, I'm going to continue the conversation in character. 
> 
> As Alice, I might be looking for some information on baking a cake or need help fixing a computer issue. Or perhaps I'm in a mood for a good book recommendation or just a friendly chat. Is there anything particular you'd want to know or discuss with Alice?
> ```

@umermansoor The output log you provided above was printed by agent Charlie in  _process_received_message function (conversable_agent.py, line 791)

```
    def _process_received_message(self, message: Union[Dict, str], sender: Agent, silent: bool):
        ...
        if not self._is_silent(silent):
            self._print_received_message(message, sender)
```

To avoid confusing and leave the log format as it was, _process_received_message function need to use the sender final silent value for logging received message or not. I will commit for this codes like:
```
def _process_received_message(self, message: Union[Dict, str], sender: Agent, silent: bool):
        ...
        if not sender._is_silent(silent):
            self._print_received_message(message, sender)
```

**Comment by wenngong** [comment]
> @umermansoor The output log you provided above was printed by agent Charlie in _process_received_message function (conversable_agent.py, line 791)

@umermansoor  done in new commit, please review, thanks!

**Review by thinkall — COMMENTED** [review]
Thank you very much for the updates. Could you also update below agents:
class UserProxyAgent(ConversableAgent)
image
class SocietyOfMindAgent(ConversableAgent)
class WebSurferAgent(ConversableAgent)

**Comment by wenngong** [comment]
> Thank you very much for the updates. Could you also update below agents: class UserProxyAgent(ConversableAgent) image class SocietyOfMindAgent(ConversableAgent) class WebSurferAgent(ConversableAgent)

@thinkall Done, please review.

**Review by thinkall — APPROVED** [review]
Thank you, LGTM!

**Review by thinkall — COMMENTED** [review]
@Hk669 , @umermansoor , do you have further comments on this PR?

**Comment by umermansoor** [comment]
@thinkall Yep, looks good to me now. Thanks

**Review by umermansoor — APPROVED** [review]

**Pull request merged**